{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `generate_liability`\n",
    "\n",
    "### What do the $\\beta$'s represent?\n",
    "  - .\n",
    "###  What do the $h^2$'s represent?\n",
    "  - heritability (def)\n",
    "\n",
    "### Explain the shape of genos\n",
    "\n",
    "### `np.random.binomial`?\n",
    "\n",
    "### `ps` variable meaning?\n",
    "\n",
    "### expr_prs meaning?\n",
    "\n",
    "### liability meaning? - How is liability defined?\n",
    "- liability threshold model; probit model / probit regression\n",
    "- map continuous sums into binary labels\n",
    "- the linear part of the generalize linear model\n",
    "- Example:\n",
    "  - label is whether indiv has disease or not (0 or 1)\n",
    "  - crossing the threshold determines whether the prediction is 0 or 1\n",
    "- Why use liability over log reg model?\n",
    "  - interpretable odds ratio\n",
    "### `np.random.normal`?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy.stats import norm\n",
    "\n",
    "def generate_liability(n_indivs, ps, \\\n",
    "                       beta1, beta2, beta3, \\\n",
    "                       beta1_h2, beta2_h2, beta3_h2):\n",
    "    \"\"\"[summary]\n",
    "\n",
    "    Args:\n",
    "        n_indivs (int): number of individuals\n",
    "        ps ([type]): [description]\n",
    "        beta2 ([type]): [description]\n",
    "        beta3 ([type]): [description]\n",
    "        beta2_h2 ([type]): [description]\n",
    "        beta3_h2 ([type]): [description]\n",
    "\n",
    "    Returns:\n",
    "        [type]: [description]\n",
    "    \"\"\"\n",
    "    n_snps = beta1.shape[0]\n",
    "    n_genes = beta1.shape[1]\n",
    "    genos = np.empty((n_indivs,n_snps)) # genotypes\n",
    "    for i in range(n_snps):\n",
    "        # sample genotypes from a binomial distribution\n",
    "        genos[:,i] = np.random.binomial(2, ps[i], size=n_indivs)\n",
    "        # rescale genotypes to have mean 0 and std 1\n",
    "        genos[:,i] = (genos[:,i] - 2*ps[i]) / np.sqrt(2*ps[i]*(1-ps[i]))\n",
    "\n",
    "    exprs = np.empty((n_indivs,n_genes))\n",
    "    expr_prs = np.dot(genos, beta1)\n",
    "    for i in range(n_genes):\n",
    "        # gene expression is standard normally distributed with mean==PRS\n",
    "        exprs[:,i] = expr_prs[:,i] + \\\n",
    "                          np.random.normal(loc=0, \n",
    "                                           scale=np.sqrt(1-beta1_h2),\n",
    "                                           size=n_indivs)\n",
    "    # the liability is thresholded to generate cases and controls\n",
    "    liability = np.dot(exprs, beta2) + np.dot(genos, beta3) + \\\n",
    "                     np.random.normal(loc=0,\n",
    "                                      scale=np.sqrt(1-beta2_h2-beta3_h2),\n",
    "                                      size=n_indivs)\n",
    "    return genos, exprs, liability\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `generate_case_control`\n",
    "- function of threshold arg?\n",
    "- `/=` operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_case_control(n_indivs, ps,\n",
    "                          beta1, beta2, beta3,\n",
    "                          beta1_h2, beta2_h2, beta3_h2,\n",
    "                          thresh, status=\"control\"):\n",
    "    \"\"\"[summary]\n",
    "\n",
    "    Args:\n",
    "        n_indivs ([type]): [description]\n",
    "        ps ([type]): [description]\n",
    "        beta1 ([type]): [description]\n",
    "        beta2 ([type]): [description]\n",
    "        beta3 ([type]): [description]\n",
    "        beta1_h2 ([type]): [description]\n",
    "        beta2_h2 ([type]): [description]\n",
    "        beta3_h2 ([type]): [description]\n",
    "        thresh ([type]): [description]\n",
    "        status (str, optional): [description]. Defaults to \"control\".\n",
    "\n",
    "    Returns:\n",
    "        [type]: [description]\n",
    "    \"\"\"\n",
    "    assert status==\"case\" or status==\"control\"\n",
    "    n_snps = beta1.shape[0]\n",
    "    n_genes = beta1.shape[1]\n",
    "    genos = np.empty((n_indivs, n_snps))\n",
    "    exprs = np.empty((n_indivs, n_genes))\n",
    "    batch = 10000\n",
    "    n_counted = 0\n",
    "\n",
    "    while n_counted < n_indivs:\n",
    "        \"\"\"\n",
    "        Generate a batch of random individuals.\n",
    "        Depending on whether cases or controls are desired, select\n",
    "        those whose liability falls on the right side of the threshold\n",
    "        \"\"\"\n",
    "\n",
    "        genos_batch, exprs_batch, liab_batch = generate_liability(\n",
    "            batch,ps,beta1,beta2,beta3,\n",
    "            beta1_h2,beta2_h2,beta3_h2)\n",
    "        if status==\"control\":\n",
    "            keep_idxs = np.where(liab_batch < thresh)[0]\n",
    "        elif status==\"case\":\n",
    "            keep_idxs = np.where(liab_batch >= thresh)[0]\n",
    "        keep_num = min(len(keep_idxs), n_indivs - n_counted)\n",
    "        keep_idxs = keep_idxs[:keep_num]\n",
    "        genos_batch = genos_batch[keep_idxs,:]\n",
    "        exprs_batch = exprs_batch[keep_idxs,:]\n",
    "        genos[n_counted:n_counted+keep_num,:] = genos_batch\n",
    "        exprs[n_counted:n_counted+keep_num,:] = exprs_batch\n",
    "        n_counted += keep_num\n",
    "        # print(f\"generating {status:s}: {n_counted:s}/{n_indivs:s}\")\n",
    "    return genos, exprs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From `scipy.stats import norm`\n",
    "- `norm.ppf`? \n",
    "### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate(n_snps, n_genes, n_geno_samples, n_expr_samples, \\\n",
    "             beta1_h2, beta2_h2, beta3_h2, prevalence=0.01):\n",
    "    \"\"\"Simulate genotype, gene expression, and case/control status according to \n",
    "    a 2-layer linear model.\n",
    "\n",
    "    SNPs --> Genes --> Trait\n",
    "    SNP-Gene effects follow an OLS\n",
    "    Gene-Trait effects follow a liability threshold model\n",
    "    \n",
    "    Args:\n",
    "        n_snps: number of SNPs (genotype variants)\n",
    "        n_genes: number of genes\n",
    "        n_geno_samples: number of individuals with only genotype info available\n",
    "        n_expr_samples: number of individuals with both genotype and \n",
    "            expression available\n",
    "        beta1_h2: variance explained (R^2) of SNP-Gene linear model\n",
    "        beta2_h2: variance explained of Gene-Trait liability model\n",
    "        beta3_h2: variance explained of direct SNP-Trait liability model\n",
    "            (not mediated by genes); Typically, we set this to 0.\n",
    "        prevalence (float, optional): Proportion or percentage reflecting how \n",
    "            common the disease is in the population. Between 0.0 and 1.0. \n",
    "            Defaults to 0.1.\n",
    "    \n",
    "    Returns:\n",
    "        genotype, expression, and labels; 'genos_plus' is the subset which has \n",
    "            gene expression data available. 'genos_only' has only genotype data.\n",
    "    \"\"\"\n",
    "    thresh = norm.ppf(1-0.01)\n",
    "    ps = np.array([0.5]*n_snps)\n",
    "\n",
    "    # randomly sample effect sizes, the betas\n",
    "    beta1 = np.random.normal(loc=0.0, scale=1.0, size=(n_snps, n_genes))\n",
    "    beta2 = np.random.normal(loc=0.0, scale=1.0, size=n_genes)\n",
    "    beta3 = np.random.normal(loc=0.0, scale=1.0, size=n_snps)\n",
    "\n",
    "    # re-scale effect sizes to set appropriate variance explained\n",
    "    # NOTE: genotypes are scaled to have mean 0 and std 1, so ps_var should be 1\n",
    "    # ps_var = 2.0*np.multiply(ps, 1-ps)\n",
    "    ps_var = np.array([1]*n_snps)\n",
    "    # beta1\n",
    "    for i in range(n_genes):\n",
    "        unscaled_varexp = np.sum(np.multiply(np.square(beta1[:,i]), ps_var))\n",
    "        scale_factor = np.sqrt(unscaled_varexp / beta1_h2)\n",
    "        beta1[:,i] /= scale_factor\n",
    "\n",
    "    # beta2\n",
    "    unscaled_varexp = np.sum(np.square(beta2))\n",
    "    scale_factor = np.sqrt(unscaled_varexp / beta2_h2)\n",
    "    beta2 /= scale_factor\n",
    "\n",
    "    # beta3\n",
    "    if beta3_h2 == 0:\n",
    "        beta3 = np.zeros(n_snps)\n",
    "    else:\n",
    "        unscaled_varexp = np.sum(np.multiply(np.square(beta3), ps_var))\n",
    "        scale_factor = np.sqrt(unscaled_varexp / beta3_h2)\n",
    "        beta3 /= scale_factor\n",
    "\n",
    "    # simulate all indivs\n",
    "    genos_only_cont, exprs_only_cont = (\n",
    "        generate_case_control(int(n_geno_samples/2), ps,\n",
    "                              beta1, beta2, beta3,\n",
    "                              beta1_h2, beta2_h2, beta3_h2,\n",
    "                              thresh, status=\"control\"))\n",
    "    genos_only_case, exprs_only_case = (\n",
    "        generate_case_control(int(n_geno_samples/2), ps,\n",
    "                              beta1, beta2, beta3,\n",
    "                              beta1_h2, beta2_h2, beta3_h2,\n",
    "                              thresh, status=\"case\"))\n",
    "    genos_plus_cont, exprs_plus_cont = (\n",
    "        generate_case_control(int(n_expr_samples/2), ps,\n",
    "                              beta1, beta2, beta3,\n",
    "                              beta1_h2, beta2_h2, beta3_h2,\n",
    "                              thresh, status=\"control\"))\n",
    "    genos_plus_case, exprs_plus_case = (\n",
    "        generate_case_control(int(n_expr_samples/2), ps,\n",
    "                              beta1, beta2, beta3,\n",
    "                              beta1_h2, beta2_h2, beta3_h2,\n",
    "                              thresh, status=\"case\"))\n",
    "    genos_only = np.concatenate((genos_only_cont, genos_only_case), axis=0)\n",
    "    phenos_only = np.array([0]*genos_only_cont.shape[0] \n",
    "                            + [1]*genos_only_case.shape[0])\n",
    "    genos_plus = np.concatenate((genos_plus_cont, genos_plus_case), axis=0)\n",
    "    exprs_plus = np.concatenate((exprs_plus_cont, exprs_plus_case), axis=0)\n",
    "    phenos_plus = np.array([0]*exprs_plus_cont.shape[0]\n",
    "                            + [1]*exprs_plus_case.shape[0])\n",
    "    return genos_only, phenos_only, genos_plus, exprs_plus, phenos_plus, \\\n",
    "        (beta1, beta2, beta3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_snps = 100\n",
    "n_genes = 10\n",
    "n_geno_samples = 5000\n",
    "n_expr_samples = 500\n",
    "beta1_h2 = 0.1\n",
    "beta2_h2 = 0.01\n",
    "beta3_h2 = 0.0\n",
    "print(simulate(n_snps, n_genes, n_geno_samples, n_expr_samples, \\\n",
    "             beta1_h2, beta2_h2, beta3_h2, prevalence=0.01))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (ds_env)",
   "language": "python",
   "name": "ds_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}